### PR TITLE
[Fix] multi-directory grouped updates sometimes create ungrouped PRs

### PR DESCRIPTION
**Pull request by Nishnha**
### What are you trying to accomplish?

Depends on https://github.com/dependabot/dependabot-core/pull/9988
This PR is stacked on top of https://github.com/dependabot/dependabot-core/pull/9988

Multi-directory grouped updates sometimes create ungrouped PRs because the [#run_ungrouped_dependency_updates](https://github.com/dependabot/dependabot-core/blob/main/updater/lib/dependabot/updater/operations/group_update_all_versions.rb#L130-L154) method relied on some old assumptions that have changed.

`#run_ungrouped_dependency_updates` makes use of `DependencySnapshot#ungrouped_dependencies` to check which dependencies do not belong to a group and require an individual PR. 
This used to work, but we recently changed the way we calculate if a dependency belongs to a group [with the introduction of multi-directory updates](https://github.com/dependabot/dependabot-core/pull/8963). 

When Dependabot could only update 1 directory at a time, marking a dependency as "handled" with `DependencySnapshot#add_handled_dependencies` meant that it had been processed.

When we added multi-directory updates, we kept track of `handled_dependencies` per directory in case we handle a dependency in one directory and don't skip over it in a later directory. 

This meant that when we changed the `current_directory` in the `DirectorySnapshot`, we lost the list of `handled_dependencies`, so dependencies belonging to an existing pull request group are not considered handled and are included in the `ungrouped_dependencies` list.

To account for this, this PR introduces `DependencySnapshot#add_handled_group_dependencies` and `DependencySnapshot#handled_group_dependencies` to mark every dependency in a group as handled across all directories and retrieve them.

### Anything you want to highlight for special attention from reviewers?

This work depends on a few [other](https://github.com/dependabot/dependabot-core/pull/9982) PRs which also share the `:dependency_has_directory` feature flag. These other PRs add in a `directory` field to `Dependabot::Dependency` that this PR uses to track which dependencies are handled.

### How will you know you've accomplished your goal?

We have a silent ecosystem test in this PR. This is also feature flagged so we can test it out on our test repo.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: updater/lib/dependabot/dependency_snapshot.rb [hunk 1]
```diff
@@ -112,13 +112,31 @@ def add_handled_dependencies(dependency_names)
       @handled_dependencies[@current_directory] = set
     end
 
+    sig { params(dependencies: T::Array[{ name: T.nilable(String), directory: T.nilable(String) }]).void }
+    def add_handled_group_dependencies(dependencies)
+      raise "Current directory not set" if @current_directory == ""
```
Comment (landongrindheim) [inline updater/lib/dependabot/dependency_snapshot.rb:117]: 🔎 We're not using `@current_directory` in this method. Do we need this guard clause?

File: updater/lib/dependabot/dependency_snapshot.rb [hunk 2]
```diff
@@ -112,13 +112,31 @@ def add_handled_dependencies(dependency_names)
...
     def handled_dependencies
       raise "Current directory not set" if @current_directory == ""
 
       T.must(@handled_dependencies[@current_directory])
     end
 
+    # rubocop:disable Performance/Sum
+    sig { returns(T::Set[String]) }
+    def handled_group_dependencies
+      T.must(@handled_dependencies.values.reduce(&:+))
```
Comment (landongrindheim) [inline updater/lib/dependabot/dependency_snapshot.rb:136]: 👍 Good call just disabling this. Seems like a bug in Rubocop.

File: updater/lib/dependabot/updater/group_update_creation.rb [hunk 3]
```diff
@@ -446,14 +467,20 @@ def deduce_updated_dependency(dependency, original_dependency)
...
           previous_requirements: original_dependency.requirements,
           package_manager: dependency.package_manager
-        )
+        }
+
+        if Dependabot::Experiments.enabled?(:dependency_has_directory)
+          dependency_params[:directory] = dependency.directory
+        end
+
+        Dependabot::Dependency.new(**dependency_params)
```
Comment (landongrindheim) [inline updater/lib/dependabot/updater/group_update_creation.rb:483]: ℹ️ Not a big deal, but we _could_ use the previous initialization pattern with

```
directory: dependency.directory if Dependabot::Experiments.enabled?(:dependency_has_directory)
```

File: updater/lib/dependabot/updater/operations/group_update_all_versions.rb [hunk 4]
```diff
@@ -99,8 +99,9 @@ def perform
         sig { returns(Dependabot::Updater::ErrorHandler) }
         attr_reader :error_handler
 
+        # rubocop:disable Metrics/AbcSize, Metrics/PerceivedComplexity
```
Comment (landongrindheim) [inline updater/lib/dependabot/updater/operations/group_update_all_versions.rb:102]: I think these are generally good cops to pay attention to, but I suspect the violations are due to the feature flagging. As such, I think it makes sense to temporarily ignore them.

**Review by landongrindheim — APPROVED** [review]

File: updater/lib/dependabot/dependency_snapshot.rb (hunk 1 above)
Comment (Nishnha) [reply updater/lib/dependabot/dependency_snapshot.rb:117]: I'm not sure if we need it -

`@current_directory` is referenced by the `DirectorySnapshot#dependency_files` and `DependencySnapshot#dependencies` methods, but I'm not sure how those are used.

If we remove the guard from here then it won't be called when `:dependency_has_directory` is enabled, so I'm erring on the side of leaving it in.

File: updater/lib/dependabot/updater/group_update_creation.rb (hunk 3 above)
Comment (Nishnha) [reply updater/lib/dependabot/updater/group_update_creation.rb:483]: Will do

Comment (Nishnha) [reply updater/lib/dependabot/updater/group_update_creation.rb]: Hmm both rubocop and sorbet are raising errors for this pattern ☹️

Comment (Nishnha) [reply updater/lib/dependabot/updater/group_update_creation.rb]: I'm going to not follow this pattern for now since it's raising errors.

**Pull request merged**